File: zara-scraper/browser/nodejs/zara.mjs
// Zara scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Surfaces:
//   - scrapeProduct(urls)         -> array of Product dicts (PDP fields from JSON-LD + DOM)
//   - scrapeSearch(url, maxPages) -> array of SearchResult dicts (PLP cards)
//
// Zara.com is Cloudflare-fronted; Scrapeless's residential fingerprinting handles
// the JS challenge transparently.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 2 } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const session = await client().browser.create({ proxyCountry, sessionTTL: DEFAULT_SESSION_TTL });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint: session.browserWSEndpoint });
      const page = await browser.newPage();
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 });
      if (readySelector) {
        try { await page.waitForSelector(readySelector, { timeout: 20000 }); } catch (_) {}
      }
      const html = await page.content();
      if (html && html.toLowerCase().includes("<html")) return html;
      lastError = new Error("empty HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

// ---------- JSON-LD helpers ----------

function* iterJsonldNodes($) {
  const blocks = $('script[type="application/ld+json"]')
    .map((_, el) => $(el).contents().text())
    .get();
  for (const raw of blocks) {
    if (!raw || !raw.trim()) continue;
    let data;
    try { data = JSON.parse(raw); } catch { continue; }
    const nodes = Array.isArray(data) ? data : [data];
    for (const node of nodes) {
      if (!node || typeof node !== "object") continue;
      if (Array.isArray(node["@graph"])) {
        for (const sub of node["@graph"]) {
          if (sub && typeof sub === "object") yield sub;
        }
      } else {
        yield node;
      }
    }
  }
}

function typeMatches(node, wanted) {
  const t = node["@type"];
  if (typeof t === "string") return t === wanted;
  if (Array.isArray(t)) return t.includes(wanted);
  return false;
}

function firstOffer(node) {
  const offers = node?.offers;
  if (Array.isArray(offers) && offers.length) return typeof offers[0] === "object" ? offers[0] : {};
  if (offers && typeof offers === "object") {
    if (Array.isArray(offers.offers) && offers.offers.length) return typeof offers.offers[0] === "object" ? offers.offers[0] : {};
    return offers;
  }
  return {};
}

function clean(value) {
  if (value == null) return null;
  const v = String(value).replace(/\s+/g, " ").trim();
  return v || null;
}

const PRODUCT_ID_RE = /-p(\d+)\.html/i;

function extractProductId(url) {
  if (!url) return "";
  const m = url.match(PRODUCT_ID_RE);
  if (m) return m[1];
  try {
    const u = new URL(url);
    return u.pathname.replace(/\/$/, "").split("/").pop().replace(/\.html$/i, "");
  } catch { return ""; }
}

// ---------- parsers ----------

export function parseProduct(html, url) {
  const $ = cheerio.load(html);

  let ld = {};
  for (const node of iterJsonldNodes($)) {
    if (typeMatches(node, "Product")) { ld = node; break; }
  }

  const offer = firstOffer(ld);

  let images = [];
  if (typeof ld.image === "string") images = [ld.image];
  else if (Array.isArray(ld.image)) images = ld.image.filter(Boolean).map(String);
  else images = $('meta[property="og:image"]').map((_, el) => $(el).attr("content")).get().filter(Boolean);

  const name = ld.name
    ? clean(ld.name)
    : clean($('meta[property="og:title"]').attr("content")) || clean($("h1").first().text());

  const description = ld.description
    ? clean(ld.description)
    : clean($('meta[name="description"]').attr("content"));

  const sku = ld.sku || ld.productID || extractProductId(url);

  let brand = "ZARA";
  if (ld.brand) {
    brand = typeof ld.brand === "object" ? (ld.brand.name || "ZARA") : String(ld.brand);
  }

  let rawPrice = offer.price ?? clean($('meta[property="product:price:amount"]').attr("content"));
  let priceValue = null;
  if (rawPrice != null) {
    const n = parseFloat(String(rawPrice).replace(/,/g, ""));
    priceValue = Number.isFinite(n) ? n : null;
  }
  let currency = offer.priceCurrency ?? clean($('meta[property="product:price:currency"]').attr("content"));

  let availability = offer.availability ?? null;
  if (typeof availability === "string" && availability.includes("/")) {
    availability = availability.split("/").pop();
  }

  let color = ld.color ?? null;
  if (!color) {
    color = clean($('[data-qa-qualifier="product-detail-color"]').first().text()
                  || $('p.product-detail-info__color').first().text());
  }

  return {
    id: sku ? String(sku) : "",
    url,
    name: clean(name) || "",
    brand: clean(brand) || "ZARA",
    description,
    price: priceValue,
    priceCurrency: currency,
    availability,
    images,
    color: typeof color === "string" ? clean(color) : null,
    category: ld.category ? clean(ld.category) : null,
  };
}

export function parseSearch(html) {
  const $ = cheerio.load(html);

  // Build a name -> product URL map from DOM anchors, since Zara's ItemList JSON-LD
  // omits per-item URLs.
  const urlByName = new Map();
  $('a[href*="-p0"][href$=".html"], a[href*="-p1"][href$=".html"]').each((_, el) => {
    const $a = $(el);
    const href = $a.attr("href") || "";
    if (!href) return;
    let absolute = "";
    try { absolute = new URL(href, "https://www.zara.com").toString(); } catch { return; }
    const label = clean($a.attr("aria-label")) || clean($a.text());
    if (label && !urlByName.has(label.toUpperCase())) {
      urlByName.set(label.toUpperCase(), absolute);
    }
    // Also index by the image basename (product id appears in image URLs)
    const img = $a.find("img").attr("src") || $a.find("img").attr("data-src");
    if (img) {
      const m = img.match(/\/(\d{8,11})/);
      if (m) {
        const pid8 = m[1].slice(0, 8);
        if (!urlByName.has(`PID:${pid8}`)) urlByName.set(`PID:${pid8}`, absolute);
      }
    }
  });

  const items = [];

  for (const node of iterJsonldNodes($)) {
    if (!typeMatches(node, "ItemList")) continue;
    for (const el of node.itemListElement || []) {
      if (!el || typeof el !== "object") continue;
      const item = (el.item && typeof el.item === "object") ? el.item : el;
      const offer = firstOffer(item);
      const name = clean(item.name) || "";
      // Zara now ships per-item URL inside the offer (`offer.url`); fall back to
      // the legacy `item.url` / `el.url` slots, and finally to a DOM lookup.
      let url = offer?.url || item.url || el.url || "";
      let image = item.image;
      if (Array.isArray(image) && image.length) image = String(image[0]);
      else if (typeof image !== "string") image = null;

      // Look up missing URL via DOM index
      if (!url && name) url = urlByName.get(name.toUpperCase()) || "";
      if (!url && image) {
        const m = image.match(/\/(\d{8,11})/);
        if (m) url = urlByName.get(`PID:${m[1].slice(0, 8)}`) || "";
      }

      const sku = item.sku || item.productID || extractProductId(url);
      let priceValue = null;
      if (offer.price != null) {
        const n = parseFloat(String(offer.price).replace(/,/g, ""));
        priceValue = Number.isFinite(n) ? n : null;
      }
      items.push({
        id: sku ? String(sku) : "",
        url,
        name,
        image,
        price: priceValue,
        priceCurrency: offer.priceCurrency ?? null,
      });
    }
    if (items.length) break;
  }

  if (!items.length) {
    const seen = new Set();
    $('a[href*="-p"][href$=".html"]').each((_, el) => {
      const $a = $(el);
      const href = $a.attr("href") || "";
      let absolute = "";
      try { absolute = new URL(href, "https://www.zara.com").toString(); } catch {}
      const sku = extractProductId(absolute);
      if (!sku || seen.has(sku)) return;
      seen.add(sku);
      const name = clean($a.attr("aria-label") || $a.find("h2,h3,span").first().text());
      const img = $a.find("img").attr("src") || $a.find("img").attr("data-src") || null;
      items.push({
        id: sku,
        url: absolute,
        name: name || "",
        image: img,
        price: null,
        priceCurrency: null,
      });
    });
  }

  return { results: items };
}

// ---------- URL helpers ----------

function addQuery(url, params) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, String(v));
  return u.toString();
}

// ---------- scrape functions ----------

export async function scrapeProduct(urls) {
  const list = Array.isArray(urls) ? urls : [urls];
  const out = [];
  for (const url of list) {
    const html = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
    out.push(parseProduct(html, url));
  }
  return out;
}

export async function scrapeSearch(url, maxPages = 1) {
  const firstHtml = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
  const out = [...parseSearch(firstHtml).results];
  for (let page = 2; page <= maxPages; page++) {
    const pageUrl = addQuery(url, { page });
    const html = await fetchRenderedHtml(pageUrl, 'script[type="application/ld+json"]');
    out.push(...parseSearch(html).results);
  }
  return out;
}

File: zara-scraper/browser/python/zara.py
"""Zara scraper using the official Scrapeless Python SDK + Playwright over CDP.

Surfaces:
- scrape_product(urls)          -> list of Product dicts (PDP fields from JSON-LD + DOM)
- scrape_search(url, max_pages) -> list of SearchResult dicts (PLP / category cards)

Zara.com is Cloudflare-fronted; Scrapeless's residential fingerprinting handles the JS
challenge transparently. We wait on `script[type="application/ld+json"]` for both surfaces.
"""

from __future__ import annotations

import json
import os
import re
from typing import Any, Dict, List, Optional
from urllib.parse import urljoin, urlparse, parse_qsl, urlencode, urlunparse

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180

# ---------------------------------------------------------------------------

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()


async def _fetch_rendered_html(
    url: str,
    ready_selector: Optional[str] = None,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 2,
) -> str:
    last_error: Optional[Exception] = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page()
                await page.goto(url, wait_until="domcontentloaded", timeout=60000)
                if ready_selector:
                    try:
                        await page.wait_for_selector(ready_selector, timeout=20000)
                    except Exception as e:
                        logger.warning("wait_for_selector {!r} failed (continuing): {}", ready_selector, e)
                html = await page.content()
                if html and "<html" in html.lower():
                    return html
                last_error = RuntimeError("empty HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

# ---------------------------------------------------------------------------
# JSON-LD helpers
# ---------------------------------------------------------------------------

def _iter_jsonld_nodes(raw_blocks: List[str]):
    for raw in raw_blocks:
        if not raw or not raw.strip():
            continue
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        nodes = data if isinstance(data, list) else [data]
        for node in nodes:
            if not isinstance(node, dict):
                continue
            graph = node.get("@graph")
            if isinstance(graph, list):
                for sub in graph:
                    if isinstance(sub, dict):
                        yield sub
            else:
                yield node


def _type_matches(node: Dict[str, Any], wanted: str) -> bool:
    t = node.get("@type")
    if isinstance(t, str):
        return t == wanted
    if isinstance(t, list):
        return wanted in t
    return False


def _first_offer(node: Dict[str, Any]) -> Dict[str, Any]:
    offers = node.get("offers")
    if isinstance(offers, list) and offers:
        return offers[0] if isinstance(offers[0], dict) else {}
    if isinstance(offers, dict):
        nested = offers.get("offers")
        if isinstance(nested, list) and nested:
            return nested[0] if isinstance(nested[0], dict) else {}
        return offers
    return {}


def _clean(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    v = re.sub(r"\s+", " ", str(value)).strip()
    return v or None

# Zara product URLs end with `-p<digits>.html`; the trailing digits are the productId.
_PRODUCT_ID_RE = re.compile(r"-p(\d+)\.html", re.I)


def _extract_product_id(url: str) -> str:
    m = _PRODUCT_ID_RE.search(url or "")
    if m:
        return m.group(1)
    return urlparse(url or "").path.rstrip("/").split("/")[-1].replace(".html", "")

# ---------------------------------------------------------------------------
# Parsers
# ---------------------------------------------------------------------------

def parse_product(html: str, url: str) -> Dict[str, Any]:
    sel = Selector(text=html)
    raw_blocks = sel.css('script[type="application/ld+json"]::text').getall()

    ld: Dict[str, Any] = {}
    for node in _iter_jsonld_nodes(raw_blocks):
        if _type_matches(node, "Product"):
            ld = node
            break

    offer = _first_offer(ld)

    images_raw = ld.get("image")
    if isinstance(images_raw, str):
        images = [images_raw]
    elif isinstance(images_raw, list):
        images = [str(x) for x in images_raw if x]
    else:
        images = sel.css('meta[property="og:image"]::attr(content)').getall()

    name = ld.get("name") or _clean(sel.css('meta[property="og:title"]::attr(content)').get()) \
        or _clean(sel.css("h1::text").get())

    description = ld.get("description") or _clean(
        sel.css('meta[name="description"]::attr(content)').get()
    )

    sku = ld.get("sku") or ld.get("productID") or _extract_product_id(url)

    brand_node = ld.get("brand")
    if isinstance(brand_node, dict):
        brand = brand_node.get("name") or "ZARA"
    elif isinstance(brand_node, str):
        brand = brand_node
    else:
        brand = "ZARA"

    price_value: Optional[float] = None
    raw_price = offer.get("price") if offer else None
    if raw_price is None:
        raw_price = _clean(sel.css('meta[property="product:price:amount"]::attr(content)').get())
    if raw_price is not None:
        try:
            price_value = float(str(raw_price).replace(",", ""))
        except (TypeError, ValueError):
            price_value = None

    currency = offer.get("priceCurrency") if offer else None
    if not currency:
        currency = _clean(sel.css('meta[property="product:price:currency"]::attr(content)').get())

    availability = offer.get("availability") if offer else None
    if isinstance(availability, str) and "/" in availability:
        availability = availability.rsplit("/", 1)[-1]

    color = ld.get("color")
    if not color:
        color = _clean(sel.css('[data-qa-qualifier="product-detail-color"] *::text').get()
                       or sel.css('p.product-detail-info__color::text').get())

    return {
        "id": str(sku) if sku else "",
        "url": url,
        "name": _clean(name) or "",
        "brand": _clean(brand) or "ZARA",
        "description": _clean(description),
        "price": price_value,
        "priceCurrency": currency,
        "availability": availability,
        "images": images,
        "color": _clean(color) if isinstance(color, str) else None,
        "category": _clean(ld.get("category")) if ld.get("category") else None,
    }


def parse_search(html: str) -> Dict[str, Any]:
    sel = Selector(text=html)
    raw_blocks = sel.css('script[type="application/ld+json"]::text').getall()

    # Build a name/pid -> product URL map from DOM anchors, since Zara's ItemList
    # JSON-LD omits per-item URLs.
    url_by_key: Dict[str, str] = {}
    for a in sel.css('a[href*="-p0"][href$=".html"], a[href*="-p1"][href$=".html"]'):
        href = a.attrib.get("href", "")
        if not href:
            continue
        absolute = urljoin("https://www.zara.com", href)
        label_raw = a.attrib.get("aria-label") or " ".join(a.css("*::text").getall())
        label = _clean(label_raw)
        if label and label.upper() not in url_by_key:
            url_by_key[label.upper()] = absolute
        img = a.css("img::attr(src)").get() or a.css("img::attr(data-src)").get()
        if img:
            m = re.search(r"/(\d{8,11})", img)
            if m:
                pid8 = m.group(1)[:8]
                key = f"PID:{pid8}"
                if key not in url_by_key:
                    url_by_key[key] = absolute

    items: List[Dict[str, Any]] = []

    for node in _iter_jsonld_nodes(raw_blocks):
        if not _type_matches(node, "ItemList"):
            continue
        for el in node.get("itemListElement") or []:
            if not isinstance(el, dict):
                continue
            item = el.get("item") if isinstance(el.get("item"), dict) else el
            offer = _first_offer(item)
            name = _clean(item.get("name")) or ""
            # Zara now ships per-item URL inside the offer; fall back to legacy slots.
            url = (offer.get("url") if isinstance(offer, dict) else None) or item.get("url") or el.get("url") or ""
            image = item.get("image")
            if isinstance(image, list) and image:
                image = str(image[0])
            elif not isinstance(image, str):
                image = None
            if not url and name:
                url = url_by_key.get(name.upper(), "")
            if not url and image:
                m = re.search(r"/(\d{8,11})", image)
                if m:
                    url = url_by_key.get(f"PID:{m.group(1)[:8]}", "")
            sku = item.get("sku") or item.get("productID") or _extract_product_id(url)
            price_value = None
            raw_price = offer.get("price") if offer else None
            if raw_price is not None:
                try:
                    price_value = float(str(raw_price).replace(",", ""))
                except (TypeError, ValueError):
                    price_value = None
            items.append({
                "id": str(sku) if sku else "",
                "url": url,
                "name": name,
                "image": image,
                "price": price_value,
                "priceCurrency": offer.get("priceCurrency") if offer else None,
            })
        if items:
            break

    if not items:
        seen = set()
        for a in sel.css('a[href*="-p"][href$=".html"]'):
            href = a.attrib.get("href", "")
            absolute = urljoin("https://www.zara.com", href)
            sku = _extract_product_id(absolute)
            if not sku or sku in seen:
                continue
            seen.add(sku)
            name = _clean(a.css("::attr(aria-label)").get()
                          or a.css("h2::text, h3::text, span::text").get())
            img = a.css("img::attr(src)").get() or a.css("img::attr(data-src)").get()
            items.append({
                "id": sku,
                "url": absolute,
                "name": name or "",
                "image": img,
                "price": None,
                "priceCurrency": None,
            })

    return {"results": items}

# ---------------------------------------------------------------------------
# URL helpers
# ---------------------------------------------------------------------------

def _add_query(url: str, **params: Any) -> str:
    parsed = urlparse(url)
    q = dict(parse_qsl(parsed.query))
    q.update({k: str(v) for k, v in params.items()})
    return urlunparse(parsed._replace(query=urlencode(q)))

# ---------------------------------------------------------------------------
# Scrape functions
# ---------------------------------------------------------------------------

async def scrape_product(urls: List[str]) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    for url in urls:
        logger.info("scraping product {}", url)
        html = await _fetch_rendered_html(url, ready_selector='script[type="application/ld+json"]')
        out.append(parse_product(html, url))
    return out


async def scrape_search(url: str, max_pages: int = 1) -> List[Dict[str, Any]]:
    """Scrape a Zara PLP. Zara is mostly infinite-scroll; one fetch returns the first batch.
    Pass `?page=N` for sections that accept pagination."""
    logger.info("scraping search {}", url)
    first_html = await _fetch_rendered_html(url, ready_selector='script[type="application/ld+json"]')
    out: List[Dict[str, Any]] = list(parse_search(first_html)["results"])
    for page in range(2, max_pages + 1):
        page_url = _add_query(url, page=page)
        page_html = await _fetch_rendered_html(page_url, ready_selector='script[type="application/ld+json"]')
        out.extend(parse_search(page_html)["results"])
    return out


def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj
